Update here string syntax

MartinGC94 opened the pull request:
# PR Summary
This updates the here string syntax to support a variable amount of quote characters in the header, which determines what the footer should look like. This enables the user to write nested here strings like so:
```
Invoke-VMScript -ScriptText @''
    Set-Content C:\Temp.txt -Value @'
    <Text>
'@
''@
```

Additionally, adding 2 or more quotes to the here string header enables the user to indent the footer like this:
```
function MyFunction
{
    @''
    Hello World
    ''@
}
```
The indentation of the footer determines which column the lines start at. For example, in the previous example the content of the here string is `Hello World` and not `    Hello World` because both the footer and the text is indented by 4 spaces.  
For more details see the related RFC : https://github.com/PowerShell/PowerShell-RFC/pull/343 or the linked issue.

## PR Context
Fixes https://github.com/PowerShell/PowerShell/issues/2337

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [X] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [X] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

MartinGC94 committed:
Add new here-string syntax for single quotes

MartinGC94 committed:
Merge branch 'PowerShell:master' into NewHere-stringSyntax

MartinGC94 committed:
Add support for double quotes in new here-string and merge both methods to one.

MartinGC94 committed:
Remove invalid test and add tests for new herestring scenarios

iSazonov commented:
Since C# got this I believe we need follow C#.

MartinGC94 commented:
@iSazonov PowerShell is not C#. We can take all the good parts from C# but there's no need to take the bad parts as well. The C# docs for raw strings mention this:
> If you need to start or end a raw string literal with quote characters, use the multi-line format

why would we want to add the same limitation here if we can avoid it?

iSazonov commented:
Since this was not in C# but is now, it will raise questions for users. That is why it should be discussed.

daxian-dbw commented:
@MartinGC94 Sorry for the delay on responding. The Engine WG discussed this PR briefly in our last meeting, and we agreed that this would be a nice improvement to have. However, since it's a language change, an RFC would be required along with this PR. Can you please submit an RFC about this topic to the https://github.com/PowerShell/PowerShell-RFC repo? Thanks!

MartinGC94 committed:
Merge branch 'PowerShell:master' into NewHere-stringSyntax

MartinGC94 commented:
@daxian-dbw Done. See: https://github.com/PowerShell/PowerShell-RFC/pull/343

MartinGC94 committed:
Merge branch 'master' into NewHere-stringSyntax

MartinGC94 committed:
Merge branch 'PowerShell:master' into NewHere-stringSyntax

MartinGC94 committed:
Use multiple quotes instead of @ chars for new here string syntax.
Trim spaces based on the footer line, rather than the shortest line.
Remove single line here string support

MartinGC94 committed:
Treat lone CR as a line ending.
